istio/src/main/scala/io/buoyant/k8s/istio/mixer/MixerApiRequests.scala

```diff
@@ -0,0 +1,121 @@
+
+    mapper.readValue[Seq[String]](yaml)
+  }
+
+  def mkReportRequest(
+    responseCode: ResponseCodeIstioAttribute,
+    requestPath: RequestPathIstioAttribute,
+    targetService: TargetServiceIstioAttribute,
+    sourceLabel: SourceLabelIstioAttribute,
+    targetLabel: TargetLabelsIstioAttribute,
+    duration: ResponseDurationIstioAttribute
+  ): ReportRequest = {
+
+    val dictionary = mkDictionary(requestPath, targetService, sourceLabel, targetLabel)
+
+    val updated = mkAttributesUpdate(dictionary, Seq(responseCode, requestPath, targetService, sourceLabel, targetLabel, duration))
+    ReportRequest(attributeUpdate = Some(updated))
+  }
+
+  def mkCheckRequest(istioREquest: IstioRequest[_]) = {
+    val TODO = Some(1L)
+    val dictionary = mkDictionary(istioREquest.requestedPath, istioREquest.targetService, istioREquest.sourceLabel, istioREquest.targetLabel)
+    val requestAttributes = Seq(istioREquest.requestedPath, istioREquest.sourceLabel, istioREquest.targetLabel, istioREquest.targetService)
+    val attributesUpdate = mkAttributesUpdate(dictionary, requestAttributes)
+    CheckRequest(TODO, Some(attributesUpdate))
+  }
+
+  private def mkDictionary(
+    requestPath: RequestPathIstioAttribute,
+    targetService: TargetServiceIstioAttribute,
+    sourceLabel: SourceLabelIstioAttribute,
+    targetLabel: TargetLabelsIstioAttribute
+  ) = {
+    val targetLabelApp = targetLabel.value.getOrElse("app", "")
+    val targetLabelVersion = targetLabel.value.getOrElse("version", "")
+    val sourceLabelApp = sourceLabel.value.getOrElse("app", "")
+    val sourceLabelVersion = sourceLabel.value.getOrElse("version", "")
+
+    //TODO: when we first added Istio integration, this was required to make the integration work. We need to double-check if we can avoid sending these extra entries
+    val customWords = Seq(
+      "app",
+      "version",
+      requestPath.name,
+      targetService.name,
+      targetLabelApp,
+      targetLabelVersion,
+      sourceLabelApp,
+      sourceLabelVersion
+    ) ++ sourceLabel.value.keySet ++ targetLabel.value.keySet
+    customWords
+  }
+
+  private def mkAttributesUpdate(customAttributeNames: Seq[String], attributes: Seq[IstioAttribute[_]]) = {
+
+    // TODO: eventually we should not have to send istioGlobalDict over the wire,
+    // and instead use positive index integers
+    val dictionaryToUse = (istioGlobalDict ++ customAttributeNames).toSet.toSeq
+
+    val allStringAttributes = attributes.filter(_.attributeType == StringAttribute).map(_.asInstanceOf[IstioAttribute[String]])
+    val allInt64Attributes = attributes.filter(_.attributeType == Int64Attribute).map(_.asInstanceOf[IstioAttribute[Long]])
+    val allStringMapAttributes = attributes.filter(_.attributeType == StringMapAttribute).map(_.asInstanceOf[IstioAttribute[Map[String, String]]])
+    val allDurationAttributes = attributes.filter(_.attributeType == DurationAttribute).map(_.asInstanceOf[IstioAttribute[Duration]])
```

Comment:
Is there any way to do this w/o casting?

Reply:
I looked at this for a bit and couldn't think of anything, due to `attributes: Seq[IstioAttribute[_]]``. Did I miss something?

Reply:
Hmm. What about a function on `IstioAttribute` taking an attribute type and optionally returning that attribute if it is of the given type?

Reply:
Another potential concern here – which might be a premature optimization – is that the repeated calls to `filter` mean that we traverse the sequence of attributes 4 times. 

As an alternative, consider using mutable collections (gasp!) and traversing the sequence a single time and appending the attribute to a different collection for each attribute type? If this is in the hot path, it might be worth using mutability.

Reply:
Alternatively, if you really want to avoid mutable collections, could always do something like this
```scala
val (stringAttrOpts, int64AttrOpts, stringMapAttrOpts, durationAttrOpts) = 
  attributes.map { 
    case attr: IstioAttribute[String] => (Some(attr), None, None, None)
    case attr: IstioAttribute[Long] => (None, Some(attr), None, None)
    case attr: IstioAttribute[Map[String, String]] => (None, None, Some(attr), None)
    case attr: IstioAttribute[Duration] => (None, None, None, Some(attr))
  }.unzip
// the resultant collections are now `Seq[Option[IstioAttribute[T]]]`, which you'd then have to `.flatten`...
```
which is kind of ugly, but is purely functional!

Reply:
With the change to IstioAttributes I suggest above, this could be something like:
```
val allStringAttributes = attributes.collect { case attr@StringAttribute(_) => attr }
```

Reply:
I've made the changes @adleong mentioned. @hawkw that's a valid concern and interesting solution, but I would argue that we should rather get through with istio 0.1.x features and address performance on our 0.2 work. Istio 0.1.x doesn't focus on performance as much and I find it unlikely that this would be an actual bottleneck as opposed to, for e.g. the lack of caching in this release.